src/GuzzleHttp/Cache/MockStorageAdapter.php

```diff
@@ -83,59 +69,26 @@ public function save(RequestInterface $request, ResponseInterface $response)
             $response = $response->withoutHeader($header);
         }
 
-        file_put_contents($this->getPath($request), Psr7\str($response));
+        list($strategy) = $this->namingStrategies;
+
+        $filename = $this->getFilename($strategy->filename($request));
 
-        $response->getBody()->seek(0);
+        $fs = new Filesystem();
+        $fs->mkdir(dirname($filename));
+
+        file_put_contents($filename, Psr7\str($response));
+        $response->getBody()->rewind();
     }
 
     /**
-     * Create a fingerprint for each request.
-     *
-     * As it is for mocking (and not for real caching), ignore some
-     * characteristics like the 'User-Agent' to avoid stale cache
-     * when updating PHP or Guzzle.
+     * Prefixes the generated file path with the adapter's storage path.
      *
-     * @param RequestInterface $request
-     * @param bool             $withHost
+     * @param string $name
      *
      * @return string The path to the mock file
      */
-    public function getPath(RequestInterface $request, $withHost = true)
+    private function getFilename($name)
     {
-        $headers = $request->getHeaders();
-        foreach ($headers as $name => $values) {
-            if (in_array($name, $this->requestHeadersBlacklist)) {
-                unset($headers[$name]);
-            }
-        }
-
-        $fingerprint = md5(serialize([
-            'method' => $request->getMethod(),
-            'path' => $request->getUri()->getPath(),
-            'query' => $request->getUri()->getQuery(),
-            'user_info' => $request->getUri()->getUserInfo(),
-            'port' => $request->getUri()->getPort(),
-            'scheme' => $request->getUri()->getScheme(),
-            'headers' => $headers,
-        ]));
-
-        if (true === $withHost) {
-            $path = sprintf(
-                '%s_%s_%s____%s',
-                str_pad($request->getMethod(), 6, '_'),
-                $request->getUri()->getHost(),
-                urldecode(ltrim($request->getUri()->getPath(), '/').'-'.$request->getUri()->getQuery()),
-                $fingerprint
-            );
-        } else {
-            $path = sprintf(
-                '%s_%s____%s',
-                str_pad($request->getMethod(), 6, '_'),
-                urldecode(ltrim($request->getUri()->getPath(), '/').'-'.$request->getUri()->getQuery()),
-                $fingerprint
-            );
-        }
-
-        return $this->storagePath.'/'.preg_replace('/[^a-zA-Z0-9_+=@\-\?\.]/', '-', $path).'.txt';
```

Comment:
I suggest to dispatch an event (Symfony EventDispatcher) here in order to allow users, with a listener, to have a custom storage path based on the request ?
For example:
{ storagePath }/api1/GET/hash.txt
{ storagePath }/api1/POST/hash.txt
{ storagePath }/api2/GET/hash.txt
....